src/alert_processing/file_system_status.py

```diff
@@ -0,0 +1,65 @@
+import logging
+from datetime import datetime, timedelta, timezone
+from pathlib import Path
+from typing import Iterable
+
+import pandas as pd
+
+__all__ = [
+    "get_modified_time",
+    "check_update_time",
+]
+
+logger = logging.getLogger(__name__)
+
+def get_modified_time(
+        files: Iterable[Path],
+        skip_dir: bool = False,
+        skip_hidden: bool = True,
+) -> pd.DataFrame:
+    paths = []
+    for path in files:
+        if skip_dir and path.is_dir():
+            continue
+        if skip_hidden and path.name[0] == '.':
+            continue
+        file_info = dict(
+            stem=path.stem,
+            path=path.as_posix(),
+            modified_datetime=datetime.fromtimestamp(
+                path.stat().st_mtime,
+                tz=timezone.utc,
+            ),
+        )
+        paths.append(file_info)
+    return pd.DataFrame(paths, columns=['stem', 'path', 'modified_datetime'])
+
+
+def check_update_time(
+        dir_path: Path,
+        current_time: datetime,
+        max_age: timedelta,
+) -> bool:
+    '''Find the most recent update time for all files in dirpath,
+    and return a status boolean if we pass certain time check thresholds.
+
+    Parameters
+    ----------
+    dir_path : Path
+        Directory path to dir containing files or station sub-directories and files
+    current_time : datetime
+        Current datetime used for determine file age
+    max_age : timedelta
+        Maximum allowed age of latest modified file before returning True.
+
+    Returns
+    -------
+    status : bool
+        Result of the check. False (default) is passing, True is alert condition
+    '''
+    modified_time_df = get_modified_time(dir_path.rglob('*'))
+    if modified_time_df.empty:
+        logger.warning(f"Unable to find any files in {dir_path}")
+        return True
+    latest_age = current_time - modified_time_df['modified_datetime'].max()
+    return latest_age > max_age
```

Comment:
Reintroduce a way to avoid notifying for the same file multiple times

Reply:
The logic in the previous implementation was:
```
    # if (max_mtime < one_hour_ago) and (max_mtime > two_hours_ago):
    if max_mtime < one_hour_ago:
        status = True
```
Where all times are in epoch seconds and `True` is the alert condition.

The commented version `if (max_mtime < one_hour_ago) and (max_mtime > two_hours_ago):` was intended to alert only once. It should trigger the alert for the first missed processing run, but will not trigger again. This condition should only exist once, where the file's update time is older than one hour ago but still younger than two hours ago. I believe I used this logic for the glacio01 monitor, because that server can go down for days at a time, and we don't need emails every hour about it.

The version `if max_mtime < one_hour_ago:` was intended to trigger every hour that the file continues to fail to update. Because the AWS processing and submission to DMI is an operational data stream with downstream users that depend on reliable data, I decided to keep this alerting every hour, so it would be harder to ignore (though it can annoyingly fill your inbox!).

It is up to you and others at GEUS how often you want to be alerted. If only once, just make sure someone is responding to the alarm.